### ISSUE TITLE
Few-shot prefix a LLM to prompt it into an instruction following mode

### ISSUE THREAD

**Issue by sanagno** [LAION-AI/Open-Assistant#341]
One of the baselines presented in the InstructGPT [paper](https://arxiv.org/pdf/2203.02155.pdf) is a "properly" prompted GPT-3 model (see footnote 6 and section 3.5). Before the specified instruction/prompt by the user, a specified prefix is prepended to the user-specified instructions.

Requirements:
- A base model (eventually the same as the actual model we are going to be using).
- A trained reward model.

After some fine-tuning on the prefixes, compare the base model and the propmpted model based on the reward attained by the reward model.

**Comment by Abubakar115e** [comment]
Hi
I am a ML student at Copenhagen University and it will be a good opportunity for me too try some of the theory that I have learned. Particularly in upcoming and temporary research obtained in my time in University.
BH
Abubakar

**Comment by sanagno** [comment]
Feel free to join the discord and contact me from there :)

**Comment by Abubakar115e** [comment]
Hi I have made this python script to test these datasets from InstructGPT and the base model from current Open-Assistant base model.
dataets: https://github.com/openai/following-instructions-human-feedback/tree/main/automatic-eval-samples

import torch
from transformers import AutoModel, AutoTokenizer
from torch.distributions import Categorical
import torch.nn as nn
import torch.optim as optim
import numpy as np

if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")  

# Load the pre-trained model and tokenizer
tokenizer = AutoTokenizer.from_pretrained("roberta-base")
base_model = AutoModel.from_pretrained("roberta-base").to('cuda')

class LogisticRegression(nn.Module):
    def __init__(self, input_dim, output_dim):
        super(LogisticRegression, self).__init__()
        self.linear = nn.Linear(input_dim, output_dim)
    def forward(self, x):
        out = self.linear(x)
        return out
class PPO:
    def __init__(self, model, optimizer, clip_epsilon, num_steps):
        self.model = model
        self.optimizer = optimizer
        self.clip_epsilon = clip_epsilon
        self.num_steps = num_steps
        self.old_log_prob = None

    def select_action(self, state):
        state = torch.FloatTensor(state).unsqueeze(0).to(device)
        prob = self.model(state)
        m = Categorical(prob)
        action = m.sample()
        self.old_log_prob = m.log_prob(action)
        return action.item()

    def update(self, state, action, reward, next_state, done):
        state = torch.FloatTensor(state).unsqueeze(0).to(device)
        next_state = torch.FloatTensor(next_state).unsqueeze(0).to(device)
        action = torch.LongTensor([action]).to(device)
        reward = torch.FloatTensor([reward]).to(device)
        done = torch.FloatTensor([done]).to(device)

        prob = self.model(state)
        m = Categorical(prob)
        curr_log_prob = m.log_prob(action)
        ratio = torch.exp(curr_log_prob - self.old_log_prob)
        gamma = 0.99
        advantage = reward + gamma * self.model(next_state) * (1-done) - self.model(state)
        surr1 = ratio * advantage
        surr2 = torch.clamp(ratio, 1-self.clip_epsilon, 1+self.clip_epsilon) * advantage
        loss = -torch.min(surr1, surr2)

        self.optimizer.zero_grad()
        loss.mean().backward()
        self.optimizer.step()

# Prepare input data
X_train = # your data
y_train = # your data
X_train = tokenizer.batch_encode_plus(X_train, return_tensors='pt', padding=True, truncation=True)['input_ids']
X_train = X_train.to('cuda')
y_train = y_train.to('cuda')
input_dim = X_train.size(1)
output_dim = y_train.size(1)

# Initialize and train the reward model
reward_model = LogisticRegression(input_dim, output_dim)
reward_model.to('cuda')
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(reward_model.parameters(), lr=0.01)

for epoch in range(100):
    optimizer.zero_grad()
    outputs = reward_model(X_train)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()

# define the prefix
prefix = # your prefix 
# Prepare input data with prefixes
X_train_prefix = X_train
for i in range(X_train_prefix.shape[0]):
    X_train_prefix[i] = prefix + X_train[i]
# define the number of epochs
num_epochs = # your number of epochs
prompted_model = base_model
optimizer = optim.Adam(prompted_model.parameters())
for epoch in range(num_epochs):
    optimizer.zero_grad()
    outputs = prompted_model(X_train_prefix)
    loss = criterion(outputs, y_train)
    loss.backward()
    optimizer.step()

X_test = # your data
X_test = tokenizer.batch_encode_plus(X_test, return_tensors='pt', padding=True, truncation=True)['input_ids']
X_test = X_test.to('cuda')
base_predictions = base_model(X_test)
prompted_predictions = prompted_model(X_test)

base_reward = reward_model(base_predictions)
prompted_reward = reward_model(prompted_predictions)

if prompted_reward > base_reward:
    print("The prompted model performed better.")
else:
    print("The base model performed better.")

optimizer = optim.Adam(base_model.parameters())
criterion = nn.CrossEntropyLoss()
clip_epsilon = 0.2
num_steps = 10
agent = PPO(base_model, optimizer , criterion, clip_epsilon, num_steps)
#define the state
state = # your state
#define the action
action = # your action
#define the advantage
advantage = # your advantage
#define the old log probability
old_log_prob = # your old log probability
for i in range(num_steps):
    agent.update(state, action, advantage, old_log_prob)
    prompted_model = base_model
    optimizer = optim.Adam(prompted_model.parameters())
    for epoch in range(num_epochs):
        optimizer.zero_grad()
        outputs = prompted_model(X_train_prefix)
        loss = criterion(outputs, y_train)
        loss.backward()
        optimizer.step()

**Comment by Abubakar115e** [comment]
Sadly my GPU is not powerful enough 
[model-vs-reward.txt](https://github.com/LAION-AI/Open-Assistant/files/10689621/model-vs-reward.txt)

**Comment by sanagno** [comment]
Thanks for the effort @Abubakar115e. The idea of this issue is to experiment with different prefixes. If you cannot run these models it would be difficult to proceed. Perhaps you can try using some quantization tricks. You should be able to run decent models for inference even on a regular GPU.

**Comment by Abubakar115e** [comment]
Yes I know that you can and use both the GPU (RTX 2070 max q) and RAM, but I only have one laptop and this work could be anywhere from a few hours to several days. I have corrected the code and should experiment with different prefixes. 
[model-vs-reward.txt](https://github.com/LAION-AI/Open-Assistant/files/10714760/model-vs-reward.txt)

**Comment by wangrui6** [comment]
Hi, I should be able to help with a AWS EC2 G5 instance with 24GB RAM. Is that enough? @sanagno

**Comment by sanagno** [comment]
That's me: Sotiris#3996 :)
